### ISSUE TITLE
How are velocities handled in BioSimSpace?

### ISSUE THREAD

**Issue by msuruzhon** [michellab/BioSimSpace#247]
Hi,

I have been running into some AMBER errors during testing, which were related to AMBER expecting velocities that were not in the input rst7 file. Since I am creating these in a custom way, I assume I have not copied the velocities over properly from the previous crd file. However, I can't see a `velocities` attribute in the molecule `propertyKeys()`. How do I store velocities properly so that they can be output in an rst7 file?

Many thanks.

**Comment by lohedges** [comment]
Could you provide some example input files including velocities? It looks like the `Sire.IO.AmberRst7` parser does support reading/writing velocities, which _should_ be mapped to a molecule property called `velocity`.

Within BioSimSpace we don't yet have full support for reading velocities from file. (I added support with NAMD when I started on the project.) Sire does provide support for generating velocities itself, so this could be used if we needed to set up a system with a particular thermal distribution of velocities.

Cheers.

**Comment by msuruzhon** [comment]
Thanks, I just realised that I can see a `velocity` property when I read an `RST7` file. However, there don't seem to be any velocity-related properties (e.g. `velocity0` and `velocity1`) in the perturbable molecules. Are these recognised by various output parsers (i.e. should I try to create them based on the "squashed" topology read by the regular `RST7` parser)? If not, I guess it makes sense to put velocity generation always on by default in the MD protocols.

**Comment by lohedges** [comment]
The merge code _should_ set the properties for the two end states based on the properties in the original molecules. However, you only get properties named `prop0` and `prop1` if the property is present in both molecules, i.e. it is a _shared_ property. With velocities, perhaps you only have this for the lambda=0 state there will only be a single `velocity` property. I'm actually not even sure what would happen in this case, e.g. if one state contained more atoms) then there would be "missing" properties in the other end state.

You could try assigning the original velocity property to your squashed topology, making sure to assign the properties to the correct atoms if they have been re-ordered.

I agree that it would be good to default to velocity generation in the configuration files. I think I do this (maybe not for all engines) when the user doesn't specify a restart simulation.

**Comment by msuruzhon** [comment]
This makes sense, many thanks for that. I will try doing this and will let you know how I get on.

**Comment by msuruzhon** [comment]
So I managed to successfully extend the functionality to velocities as well but I ran into some problems with some of the core `BioSimSpace` functionality implicitly removing some of the velocities. I narrowed it down to the `System._set_water_topology()` function. It wasn't easy to catch, so I am using a modified version of the function (basically forcing its execution even if the water naming convention is correct):

```python
from Sire import Mol as _SireMol
from Sire import IO as _SireIO
from Sire import System as _SireSystem

import BioSimSpace as BSS


def _set_water_topology_modified(self, format, property_map={}):
    """Internal function to swap the water topology to AMBER or GROMACS format.

        Parameters
        ----------

        format : string
            The format to convert to: either "AMBER" or "GROMACS".

        property_map : dict
           A dictionary that maps system "properties" to their user defined
           values. This allows the user to refer to properties with their
           own naming scheme, e.g. { "charge" : "my-charge" }
    """

    # Validate input.

    if type(format) is not str:
        raise TypeError("'format' must be of type 'str'")

    if type(property_map) is not dict:
        raise TypeError("'property_map' must be of type 'dict'")

    # Strip whitespace and convert to upper case.
    format = format.replace(" ", "").upper()

    # We allow conversion to AMBER or GROMACS format water toplogies. While
    # AMBER requires a specific topology, GROMACS can handle whatever. We
    # leave this format as an option in case we want to allow the user to
    # swap the topology in future.
    if format not in ["AMBER", "GROMACS"]:
        raise ValueError("'format' must be 'AMBER' or 'GROMACS'.")

    # Get the water molecules.
    waters = self.getWaterMolecules()

    if True:

        # Don't perform conversion if the topology already matches the
        # the template for the desired format.

        # if format == "AMBER":
        #     if waters[0].isAmberWater():
        #         return
        #
        # elif format == "GROMACS":
        #     if waters[0].isGromacsWater():
        #         return

        # There will be a "water_model" system property if this object was
        # solvated by BioSimSpace.
        if "water_model" in self._sire_object.propertyKeys():
            water_model = self._sire_object.property("water_model").toString()

        # Otherwise, convert to an appropriate topology.
        else:
            num_point = waters[0].nAtoms()

            if num_point == 3:
                # TODO: Assume TIP3P. Not sure how to detect SPC/E at present.
                water_model = "TIP3P"
            elif num_point == 4:
                water_model = "TIP4P"
            elif num_point == 5:
                water_model = "TIP5P"

        if format == "AMBER":
            new_waters = _SireIO.setAmberWater(self._sire_object.search("water"), water_model)
        else:
            new_waters = _SireIO.setGromacsWater(self._sire_object.search("water"), water_model)

        # Create a new system and molecule group.
        system = _SireSystem.System("BioSimSpace System")
        molgrp = _SireMol.MoleculeGroup("all")

        # Create a dictionary of water molecule numbers to list index.
        wat_num_to_idx = {}
        for idx, wat in enumerate(waters):
            wat_num_to_idx[wat.number()] = idx

        # Add all moleclules from the original system into the new one,
        # preserving the order.
        for mol in self.getMolecules():
            try:
                # Water molecule.
                molgrp.add(new_waters[wat_num_to_idx[mol.number()]])
            except:
                # Non-water molecule.
                molgrp.add(mol._sire_object)

        # Add the group to the system.
        system.add(molgrp)

        # Copy across system properties.
        for prop in self._sire_object.propertyKeys():
            system.setProperty(prop, self._sire_object.property(prop))

        # Set the system.
        self._sire_object = system

        # Reset the index mappings.
        self._reset_mappings()

        # Update the molecule numbers.
        self._mol_nums = self._sire_object.molNums()

system = BSS.IO.readMolecules(["equilibrate.crd", "equilibrate.prm7"])
BSS.IO.saveMolecules("before", system, "rst7,gro87")
_set_water_topology_modified(system, "AMBER")
BSS.IO.saveMolecules("after", system, "rst7,gro87")
```

The input files can be found [here](https://github.com/michellab/BioSimSpace/files/7506642/Test.zip).

It can be clearly seen from the `before` and `after` files that the water velocities have been substituted with nonsensical numbers, which is probably related to some incorrect memory handling (I think this is just because the `velocity` property is removed when calling `_SireIO.setAmberWater()` or `_SireIO.setGromacsWater()`). Is this a `Sire` issue or am I misusing the functionality somehow?

Many thanks.

**Comment by lohedges** [comment]
Hi there. This functionality is implemented behind the scenes in Sire. It works by loading known water topologies from file and swapping the existing molecules. Clearly velocities aren't being preserved in this process. (It should only replace names and overwrite any properties associated with the previous molecule.)

If you upload some example input files I can try to debug / fix this.

**Comment by msuruzhon** [comment]
Hi, many thanks for the reply, this makes sense. What type of input files would you like me to send you apart from the ones in the example ("equilibrate.crd" and "equilibrate.prm7")?

**Comment by lohedges** [comment]
Apologies, I didn't see the link when reading on my phone. I'll try with those and let you know if I need anything else.

I'm confused why always replacing the water topology would work. If anything, I'd expect replacing it to cause the issue since the water templates don't have velocities.

**Comment by msuruzhon** [comment]
No worries, I meant that I modified the function so that it is always called. It is precisely this replacement that causes this issue, as far as I can see it.

**Comment by lohedges** [comment]
I've now fixed this in Sire. The issue was that properties that were unique to the original water topology, e.g. velocities, weren't being preserved during the topology swap. Once the CI build has finished you'll need to update your local conda version of Sire to get the changes.

Thanks again for raising this.

**Comment by msuruzhon** [comment]
Many thanks for that Lester, it works for me now.

**Comment by lohedges** [comment]
Just a heads up that I've now replaced the `system._updateCoordinates` method with `system._updateCoordiantesAndVelocities`, which will copy back update velocities if they are present in the new system. I imagine that you've already done something like this for AMBER, but this should work in general. (I think you need to do some work to unsquash the system to get it back into the original format first.)

**Comment by msuruzhon** [comment]
Hi @lohedges, I am playing around with velocities again and I am seeing that they are again being zeroed for some reason. Here is a script to reproduce that:

```python
import BioSimSpace.Sandpit.Exscientia as BSS
from BioSimSpace.Sandpit.Exscientia._SireWrappers import System
from BioSimSpace.Sandpit.Exscientia.Align._squash import _squash

mol1 = BSS.Parameters.gaff("c1ccccc1C").getMolecule()
mol2 = BSS.Parameters.gaff("c1ccccc1").getMolecule()
pert_mol = BSS.Align.merge(mol1, mol2)
solv_system = BSS.Solvent.tip3p(
    molecule=pert_mol, box=3 * [3 * BSS.Units.Length.nanometer]
)
solv_system_pruned = System(solv_system[0].copy() + solv_system[-1].copy())
solv_system_pruned.setBox(*solv_system.getBox())
protocol = BSS.Protocol.Equilibration(
    timestep=BSS.Types.Time(1, "femtosecond"),
    runtime=BSS.Types.Time(1, "femtosecond"),
    temperature=BSS.Types.Temperature(10, "kelvin"),
)
process = BSS.Process.Gromacs(solv_system_pruned, protocol, ignore_warnings=True)
process.start()
process.wait()
system_with_velocities = process.getSystem()
assert system_with_velocities is not None

squashed_system, _ = _squash(system_with_velocities)
BSS.IO.saveMolecules("temp", squashed_system, "gro87,rst7")
```
As you can see, the output GRO file has all the velocities, whereas the RST7 file has all the water velocities zeroed. Is that a regression or am I doing something wrong? Many thanks.

**Comment by lohedges** [comment]
Hi there. I think what's going on is that the velocities are being lost during the water topology conversion. When saving to GROMACS it doesn't need to do anything, since they are already GROMACS format, but with AMBER it is swapping the topology. I'll check this and add a fix to copy across any molecular properties that aren't replaced during the topology swap. (I thought it already did this, though.)

Just to note, when I ran your example the `_squash` function returned a single item tuple, so I needed to pass `squashed_system[0]` to `saveMolecules`. Is this the intended behaviour, or is my version of the sandpit out of sync somehow?

**Comment by msuruzhon** [comment]
Thanks Lester. Ah yes sorry I didn't test this line, it should be 

```python
squashed_system, _ = _squash(system_with_velocities)
```

I will edit it above.

**Comment by lohedges** [comment]
Okay, I've fixed it. I did have code to check for properties that were unique to the original molecule when updating the topology, but I was using the _wrong_ molecule when checking whether we needed to add them. [Here](https://github.com/OpenBioSim/sire/commit/311d304fe072d3dd2925f786e4c320cee7334c63) is the fix. (Properties from the updated topology take precedence, but we need to add those that were present in the old molecule but aren't in the new one.)

**Comment by msuruzhon** [comment]
Many thanks @lohedges, I will give it another try after the new `sire` builds then. Will that be released as a `2023.1.4`?

**Comment by lohedges** [comment]
This will still be `2023.1.3`, but a dev build, i.e. the conda package from the `dev` label will have a build number that indicates the number of commits since the last tag. I'm just creating a PR, which should be reviewed when Christopher gets a chance. (He's away at a workshop today and tomorrow, so it might not be looked at until he's back.)
